Nothing to see here...

gcla opened the pull request.

gcla committed:
A widget that draws a hi-tech shark fin in the background...

Credit to @pocc Ross Jacobs for the concept and code ;-)

gcla committed:
Activate the shark fin when "Please wait" is on the screen.

pocc commented:
I like this, but am having trouble testing it. On my linux VM, I try to build on commit `3f9f0ea`, but it errors out:

```bash
bash$ ./termshark
./termshark: line 1: syntax error near unexpected token `newline'
./termshark: line 1: `!<arch>'
```

This is probably the same issue as the one on #78.

gcla commented:
Hi - I made a comment on issue 78: https://github.com/gcla/termshark/issues/78

pocc commented:
Assuming that if you run this on an interface that should not expect packets, you'll see an animation? I'm building in my linux VM and I see this:

```bash
rj@vmbuntu:~/go/src/github.com/gcla/termshark$ git rev-parse --short HEAD 
3f9f0ea
rj@vmbuntu:~/go/src/github.com/gcla/termshark$ git install ./...
rj@vmbuntu:~/go/src/github.com/gcla/termshark$ sudo -i
root@vmbuntu:~# /home/rj/go/bin/termshark
```

<img width="638" alt="Screen Shot 2020-01-29 at 11 59 46 PM" src="https://user-images.githubusercontent.com/10995145/73430938-10d3f080-42f4-11ea-929b-d8e59fb099b8.png">

Can you post an image of what you see?

gcla commented:
Well.. you're right. I wasn't sure when to activate it. I thought that it might be annoying when packets are loading because the UI is fully functional then. So right now, it only appears when a modal dialog appears, like the "Please wait" dialog. And that doesn't happen very often really. I was cheating by using copy-mode with a slow custom copy command - `slowcopy.sh`:

```bash
#!/usr/bin/env bash
sleep 3s
exec xsel -i -b
```

and then

```toml
[main]
  copy-command = ["/home/gcla/bin/slowcopy.sh"]
```

Then I navigate to the hex view, hit 'c', then hit ctrl-c to copy.

What's your opinion on when it should pop up?

gcla commented:
Forgot you were on macOS - that script should probably be

```bash
#!/usr/bin/env bash
sleep 3s
exec pbcopy
```

pocc commented:
I'm having trouble installing this on my linux VM. I have no problems installing termshark with `master`. I may well be doing something funky.

```bash
rj@vmbuntu:~/go/src/github.com/gcla/termshark$ go get
# github.com/gcla/termshark
./utils.go:241:15: cannot use &err (type *error) as type *gowid.KeyValueError in return argument
./utils.go:263:15: cannot use &err (type *error) as type *gowid.KeyValueError in return argument
rj@vmbuntu:~/go/src/github.com/gcla/termshark$ git branch
  master
* rossshark
rj@vmbuntu:~/go/src/github.com/gcla/termshark$ git rev-parse --short HEAD
3f9f0ea
```

If you wouldn't mind terribly, can you provide a gif of what it looks like with your config + copy setup?

gcla commented:
yep no problem - here's a link to an animation:

https://drive.google.com/open?id=15pPf1Uc_KTUlUXh4vJSSLJX_0wb89oh3

You can see that it only appears with an artificially delayed copy-command. I think I need a better way to make it show up. Maybe just an easter-egg?

gcla commented:
About the compile error - does this program run for you?

https://play.golang.org/p/OA6YWvKmTrt

If you just drop it somewhere as foo.go and then type

```
go run foo.go
```

```go
package main

type MyError struct{}

var _ error = MyError{}

func (e MyError) Error() string {
	return "my error"
}

func (e MyError) Something() {}

type Somethinger interface {
	Something()
}

func sillyAPI() (string, error) {
	return "", nil
}

func testError() error {
	worked := false
	if _, err := sillyAPI(); err == nil {
		// err scoped to here, right?
		worked = true
	}
	if !worked {
		// thought this would be a "new" err
		err := MyError{}
		var _ Somethinger = err // regular error type won't work here
		var _ error = &err
		return &err
	}
	return nil
}

func main() {
	testError()
}
```

gcla commented:
Or - in utils.go, try changing

```go
if _, err := os.Stat(tsharkBin); err == nil {
```

to 

```go
if _, err2 := os.Stat(tsharkBin); err2 == nil {
```

gcla commented:
Maybe we could make the fin pop up after a few seconds of keyboard or mouse inactivity during a live capture... And then disappear again as soon as the user hits a key

gcla commented:
You can probably tell I think this is the most important feature to ship next ;-)

pocc commented:
I get you. Paraphrasing Frank Herbert, the text must flow ;)

gcla commented:
It's been a while but I loved that book...

Easier way to trigger it - open up termshark with a biggish pcap, then select a packet and open stream reassembly

gcla commented:
I'm considering merging this for now, then worrying about optimizing it later :-) Ok with you?

pocc commented:
## Environment Difficulties

I deleted ~/go and upgraded to go1.13.3 which seems to have solved that problem. Thanks for the environment troubleshooting.

## Fin placement

> Maybe we could make the fin pop up after a few seconds of keyboard or mouse inactivity during a live capture... And then disappear again as soon as the user hits a key

This is exactly when it should appear. But I wouldn't limit this screensaver to live captures. I would use these timings:

1. Waiting, any context (incl live capture with no packets and copy mode): Start immediately, break on state change
2. No activity: Start after 60s of no activity, break on user input

## Fins with messages

```
1. Waiting, with message

00000000000000000000000000111111111
00000000000000000000011111111111110
00000000000000000111111111111111100
00000000000000111111111111111111000
00000000000011111111111111111110000
00000000001111111111111111111110000
00000000111111111111111111111100000
00000001111111111111111111111100000
00000011111111111111111111111100000
00000111111111111111111111111100000
00001111111111111111111111111100000
0001111111┏━━━━━━━━━━━━┓11111100000
0011111111┃ No Packets ┃11111100000
0111111111┗━━━━━━━━━━━━┛11111100000
01111111111111111111111111111110000
11111111111111111111111111111110000
11111111111111111111111111111111000
11111111111111111111111111111111100

00000000000000000000000000111111111
00000000000000000000011111111111110
00000000000000000111111111111111100
00000000000000111111111111111111000
00000000000011111111111111111110000
00000000001111111111111111111110000
00000000111111111111111111111100000
00000001111111111111111111111100000
00000011111111111111111111111100000
00000111111111111111111111111100000
00001111111111111111111111111100000
0001111111┏━━━━━━━━━━━━┓11111100000
0011111111┃ Loading... ┃11111100000
0111111111┗━━━━━━━━━━━━┛11111100000
01111111111111111111111111111110000
11111111111111111111111111111110000
11111111111111111111111111111111000
11111111111111111111111111111111100

2. Sharksaver

00000000000000000000000000111111111
00000000000000000000011111111111110
00000000000000000111111111111111100
00000000000000111111111111111111000
00000000000011111111111111111110000
00000000001111111111111111111110000
00000000111111111111111111111100000
00000001111111111111111111111100000
00000011111111111111111111111100000
00000111111111111111111111111100000
00001111111111111111111111111100000
0001111111┏━━━━━━━━━━━━┓11111100000
0011111111┃ Sharksaver ┃11111100000
0111111111┗━━━━━━━━━━━━┛11111100000
01111111111111111111111111111110000
11111111111111111111111111111110000
11111111111111111111111111111111000
11111111111111111111111111111111100
```

Note that each of these boxes are 14x3 so all that would need to change would be the message inside.

What do you think?

pocc commented on widgets/rossshark/rossshark.go:
```diff
@@ -0,0 +1,168 @@
+//======================================================================
+
+// Plagiarized from https://www.calhoun.io/creating-random-strings-in-go/
+const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```
Can we change the charset to match `xxd`/`tshark -x` output like so?

```golang
// Use charset [a-f0-9] to mirror tshark -x/xxd hex output 
const charset = "abcdef0123456789" 
```

gcla replied:
yep, will do

gcla committed:
Use an xxd-compatible character set suggestion - from Ross

gcla committed:
Make sure the shark fin doesn't grab all left/right keystrokes!

gcla committed:
Make the fin less aggressive

I don't know what the right interval is. Would frequent terminal
updates be annoying? For now, I'll only enable it if the terminal is
updating anyway with data from a live capture.

gcla committed:
Activate the fin after 30s of inactivity in live capture mode

gcla committed:
Add a note to the user guide on disabling the fin
